### PR TITLE
Tool recommendation in Galaxy using deep learning

### PR DESCRIPTION
**Pull request by anuprulez**
The tool recommendation system analyses the complete set of workflows available on Galaxy’s European server using a deep learning approach to predict tools. The approach learns the higher-order dependencies in the sequences of tools. An accuracy (average precision) of ~ 99% (for top-1, top-2 and top-3 predictions) is achieved. Recurrent neural network (Gated recurrent units), a variant of a neural network, is used as the deep learning architecture/network. The parameters of this network are optimised using Bayesian optimisation. The predictive system can be used in two different user interface integrations available on the Galaxy’s European server. 

**Model:**
An `hdf5` file (present at 'database/trained_model.hdf5') merges all the necessary files (including the model weights, config, tool dictionaries, usage weights) into one model file which is sufficient for recreating the trained deep learning model and making predictions. To create the model, please follow the instructions given in the following Github repository: https://github.com/anuprulez/similar_galaxy_workflow/tree/release_tool_recommendation_v_03_19

**API:**
There is an API (included in `api/workflows.py`) call to fetch the predicted tools for a tool or a sequence of tools. It takes a tool name or a sequence of tool names as an input to show the predicted tools. 

**Predictions:**
The tool predictions are shown in two places:

(a) While creating a workflow, each tool shows a `right arrow`. Clicking on this arrow opens up a modal popup which loads the predicted tools. A simple demo:

tool_prediction_demo

(b) After executing any tool, a tree is shown where the leaves are the predicted tools which can be used for further analysis.

preditions_d3_tree

**Plots:**
Precision plots showing top-1, top-2 and top-3 prediction accuracies (the approach shown in the plot (d) is used to create the recommendation system). The line plots show the mean precision over 10 runs of the same experiment and the shaded regions show one standard deviation above and below the mean.

precision_tool_prediction

(a) Deep network: The neural network with only dense layers.
(b) CNN: The neural network with convolutional and max-pooling layers (mostly used for images).
(c) RNN: The neural network with recurrent layers (mostly used for sequential data).
(d) RNN with weighted loss: The same network as in (c) but the cross-entropy loss is weighted by the tools' usage frequencies. It achieves the best accuracy with the least variance over 10 runs. 

Thank you!

### CODE REVIEW

**Comment by hexylena** [comment]
Something strange happened with the diff. I'll need to figure out what happened.

File: client/galaxy/scripts/mvc/tool-recommendation.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            let toolIdSlash = toolId.split("/");
+            toolId = toolIdSlash[toolIdSlash.length - 2];
+        }
+        Utils.request({
+            type: "POST",
+            url: `${getAppRoot()}api/workflows/get_tool_predictions`,
+            data: {"tool_sequence": toolId},
+            success: data => {
+                if (data !== null && data.predicted_data.children.length > 0) {
+                    self.$el.append("<div class='infomessagelarge'>You have used " + data.predicted_data.name + " tool. For further analysis, you may use the following tools.</div>")
```
Comment (hexylena) [inline client/galaxy/scripts/mvc/tool-recommendation.js]:
```suggestion
                    self.$el.append("<div class='infomessagelarge'>You have used " + data.predicted_data.name + " tool. For further analysis, you could try using the following tools.</div>")
```

**Comment by anuprulez** [comment]
The first 3 commits were not there when I create the PR, maybe there is some issue occurred due to the change of the base-branch

**Comment by hexylena** [comment]
No need to make changes on your side. I'll rebase / handle anything that needs to be done. I will likely merge into a -test branch first so it doesn't take down main again, until we can work out any issues.

**Comment by anuprulez** [comment]
Ok, thanks!

**Comment by bgruening** [comment]
That will make a lovely demo at GCC! :)

**Comment by hexylena** [comment]
Deploying this out on test.

What did you think of my suggestion to also include the scores, instead of just a list of tools? Would you be able to include that?

**Comment by anuprulez** [comment]
Thanks for this suggestion!

I have sorted the predicted tools in the descending order of their learned scores i.e. the first tool has the highest score, the second one has the next highest and so on. The values I get as the scores are real numbers, some are big and some are low. I am not sure if I should normalize the scores for all the predictions or just top-10 or top-20 which I show. In my opinion, it would be misleading to show those scores unless they are normalized in some way. For a user, a tool with a 60% score may be more useful than the topmost tool with a 90% score. That's why I just show the 10 or 20 topmost used tools.

**Comment by hexylena** [comment]
Sure, just thought scores might be interesting for users to see. I know I would be curious, is tool 1 WAY more highly recommended than tool 2? are tools 1-10 all relatively good suggestions, or are tools e.g. 1-2 good and 3-10 really uncertain recommendations. (Also the panel is really empty, and scores would balance it nicely / provide more information)

Could you find a normalization? 

Edit: Users of course will choose their tools that they want, no matter what order you recommend them in, I don't think you saying "90% vs 60%" recommendation will bias their choice that much, unless it's two tools that serve similar puruposes

**Comment by anuprulez** [comment]
Ok, first I will give it a try locally. thanks!

**Comment by anuprulez** [comment]
Added `% scores` for predicted tools normalized by the `maximum` predicted score. The topmost tool will have `100%` score and the rest will have scores relative to the topmost tool.

predicted_scores

d3_tree_score

**Comment by hexylena** [comment]
Love it!

**Comment by bgruening** [comment]
@anuprulez can you rebase this?

**Comment by hexylena** [comment]
@bgruening I can, don't worry. Just one commit, I'll cherry pick :)

**Comment by hexylena** [comment]
Today I finally have some bit of time, so going to make some progress on it

**Comment by hexylena** [comment]
The "server killer" in this PR seems to be the combination of keras in venv + loaded once in Galaxy **AND** `log-maxsize = 134217728` set in uwsgi.

That's fun. Fyi @natefoo for when this gets upstreamed.

Good thing all our logs go to journald, and these get cleaned through a different method, and maxsize isn't important anymore.

**Comment by hexylena** [comment]
I commented them out on a whim, earlier yesterday I'd run a flask app under uWSGI and it *would* print an error in the logs whenever the application failed to load. I figured "hey maybe it's being buffered" and just commented out logging things and got lucky.

I'm... not super motivated to debug this in detail so we're just going to chalk it up to my magnetic personality affecting the electrons and servers :)

**Comment by hexylena** [comment]
Ok due to the messy branches (no one's fault) I'm merging this into release_19.05_europe outside of this PR.

Thanks so much for doing this @anuprulez, sorry it took so long to get it in.

**Comment by anuprulez** [comment]
@erasche thanks a lot for your efforts to see it through
